MaterialLib/MPL/Properties/PermeabilityMohrCoulombFailureIndexModel.cpp

```diff
@@ -0,0 +1,155 @@
+#include "BaseLib/Error.h"
+#include "MaterialLib/MPL/Medium.h"
+#include "MaterialLib/MPL/Utils/FormEigenTensor.h"
+#include "MaterialLib/MPL/Utils/GetSymmetricTensor.h"
+#include "MathLib/KelvinVector.h"
+#include "MathLib/MathTools.h"
+#include "ParameterLib/CoordinateSystem.h"
+#include "ParameterLib/Parameter.h"
+
+namespace MaterialPropertyLib
+{
+template <int DisplacementDim>
+PermeabilityMohrCoulombFailureIndexModel<DisplacementDim>::
+    PermeabilityMohrCoulombFailureIndexModel(
+        std::string name, ParameterLib::Parameter<double> const& k0,
+        double const kr, double const b, double const c, double const phi,
+        double const k_max, double const t_sigma_max,
+        ParameterLib::CoordinateSystem const* const local_coordinate_system)
+    : k0_(k0),
+      kr_(kr),
+      b_(b),
+      c_(c),
+      phi_(boost::math::constants::degree<double>() * phi),
+      k_max_(k_max),
+      t_sigma_max_(t_sigma_max),
+      local_coordinate_system_(local_coordinate_system)
+{
+    name_ = std::move(name);
+}
+
+template <int DisplacementDim>
+void PermeabilityMohrCoulombFailureIndexModel<DisplacementDim>::checkScale()
+    const
+{
+    if (!std::holds_alternative<Medium*>(scale_))
+    {
+        OGS_FATAL(
+            "The property 'PermeabilityMohrCoulombFailureIndexModel' is "
+            "implemented on the 'medium' scale only.");
+    }
+}
+
+template <int DisplacementDim>
+PropertyDataType
+PermeabilityMohrCoulombFailureIndexModel<DisplacementDim>::value(
+    VariableArray const& variable_array,
+    ParameterLib::SpatialPosition const& pos, double const t,
+    double const /*dt*/) const
+{
+    auto const& stress_vector = std::get<SymmetricTensor<DisplacementDim>>(
+        variable_array[static_cast<int>(Variable::stress)]);
+
+    auto const& stress_tensor =
+        formEigenTensor<3>(static_cast<PropertyDataType>(stress_vector));
+
+    Eigen::SelfAdjointEigenSolver<Eigen::Matrix<double, 3, 3>>
+        eigenvalue_solver(stress_tensor);
+
+    // Principle stress
+    auto const sigma = eigenvalue_solver.eigenvalues();
+
+    auto k_data = k0_(t, pos);
+
+    double const max_sigma = std::max(std::fabs(sigma[0]), std::fabs(sigma[2]));
+
+    if (max_sigma < std::numeric_limits<double>::epsilon())
+    {
+        return fromVector(k_data);
+    }
+
+    double const sigma_m = 0.5 * (sigma[2] + sigma[0]);
+    double f = 0.0;
+    if (sigma_m > 0.5 * c_ / std::tan(phi_))
+    {
+        // von Mises failure criterion
+        double val = 0.0;
+        for (int i = 0; i < 3; i++)
+        {
+            int const i1 = i % 3;
+            int const i2 = (i + 1) % 3;
+            val += (sigma[i2] - sigma[i1]) * (sigma[i2] - sigma[i1]);
+        }
+        double const J = std::sqrt(0.5 * val);
+        f = J / t_sigma_max_;
+    }
+    else
+    {
+        // von Mohr Coulomb failure criterion
+        f = 0.5 * std::fabs(sigma[2] - sigma[0]) /
+            (c_ * std::cos(phi_) - sigma_m * std::sin(phi_));
+    }
```

Comment:
@wenqing please help me understand your changes:

1. the apex value of sigma_m is c/tan phi ... based on what logic did you decide to make the distinction at exactly 0.5 that value in L90? If there's no reason, is it a magic number?
2. f_mc (sigma_m = 0.5 c/tan(phi) = (sigma_2 - sigma_1)/(c*cos(phi)) Is it ensured that f_vonMises = f_MC at this point? They should transition continuously. Isn't t_sigma_max_ is an independent parameter? So how would continuity be guaranteed?
3. With a transition to von Mises without an actual tension cutoff, low shear but high tension will still not lead to failure.

Reply:
> @wenqing please help me understand your changes:
> 
>     1. the apex value of sigma_m is c/tan phi ... based on what logic did you decide to make the distinction at exactly 0.5 that value in L90? If there's no reason, is it a magic number?
> 
>     2. f_mc (sigma_m = 0.5 c/tan(phi) = (sigma_2 - sigma_1)/(c*cos(phi)) Is it ensured that f_vonMises = f_MC at this point? They should transition continuously. Isn't t_sigma_max_ is an independent parameter? So how would continuity be guaranteed?
> 
>     3. With a transition to von Mises without an actual tension cutoff, low shear but high tension will still not lead to failure.

Because the criterion of
```
f_shear = tau_max / (c * cos_phi - min(sigma_m, tensile_strength_parameter) * sin_phi)
```
may still have the apex point if tensile_strength_parameter=c/tan(phi). My previous idea is that if sigma_m is greater than the the half of c/tan(phi) (half distance to the apex from 0, not good with that artificial conditon), use von Mises, otherwise use the Mohr-Coulomb.  That does not fit (2) and (3) of your comments. I will provide an improvement soon taking account of  your suggestions.

Reply:
I see ... I wouldn't worry too much about this aspect: tensile_strength_parameter=c/tan(phi). We perform no parameter consistency checks in any of our models, I believe. So if people want to enter false numbers, they can. At least we provide a well defined (documented) criterion and people can choose meaningful parameters.

Reply:
GIGO principle ;-)

Reply:
@nagelt  Just updated. The changes are:
1. Input name is changed to tensile_strength_parameter.
2. if (sigma_m > tensile_strength_parameter), compute f = max(f_MC, f_vM) as you suggested. Meanwhile, f = f_vM if sigma_m = c/tan(phi).
3. if  (sigma_m <= tensile_strength_parameter), f = f_MC.